x/ecocredit/server/marketplace/update_sell_orders_test.go

```diff
@@ -0,0 +1,262 @@
+
+func TestUpdateSellOrder_AskPrice(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	testSellSetup(t, s, batchDenom, ask.Denom, ask.Denom[1:], classId, start, end, creditType)
+
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(any interface{}, p *ecocredit.Params) {
+		p.CreditTypes = []*ecocredit.CreditType{&creditType}
+	}).Times(2)
+	expiration := time.Now()
+	_, err := s.k.Sell(s.ctx, &marketplace.MsgSell{
+		Owner: s.addr.String(),
+		Orders: []*marketplace.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "5.22", AskPrice: &ask, DisableAutoRetire: false, Expiration: &expiration},
+		},
+	})
+	assert.NilError(t, err)
+
+	orderBefore, err := s.marketStore.SellOrderTable().Get(s.ctx, 1)
+	// can update price of same denom
+	askUpdate := sdk.NewInt64Coin(ask.Denom, 25)
+	_, err = s.k.UpdateSellOrders(s.ctx, &marketplace.MsgUpdateSellOrders{
+		Owner: s.addr.String(),
+		Updates: []*marketplace.MsgUpdateSellOrders_Update{
+			{SellOrderId: 1, NewAskPrice: &askUpdate},
+		},
+	})
+	assert.NilError(t, err)
+	orderAfter, err := s.marketStore.SellOrderTable().Get(s.ctx, 1)
+	assert.NilError(t, err)
+	assert.Equal(t, orderAfter.AskPrice, askUpdate.Amount.String())
+	assert.Equal(t, orderBefore.MarketId, orderAfter.MarketId)
+
+	orderBefore, err = s.marketStore.SellOrderTable().Get(s.ctx, 1)
+	assert.NilError(t, err)
+
+	askUpdate = sdk.NewInt64Coin("ubar", 18)
+	_, err = s.k.UpdateSellOrders(s.ctx, &marketplace.MsgUpdateSellOrders{
+		Owner: s.addr.String(),
+		Updates: []*marketplace.MsgUpdateSellOrders_Update{
+			{SellOrderId: 1, NewAskPrice: &askUpdate},
+		},
+	})
+	assert.NilError(t, err)
+
+	order, err := s.marketStore.SellOrderTable().Get(s.ctx, 1)
+	assert.NilError(t, err)
+
+	assert.Equal(t, order.AskPrice, askUpdate.Amount.String())
+	assert.Equal(t, order.MarketId, orderBefore.MarketId+1)
+}
+
+func TestUpdateSellOrder_Expiration(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	testSellSetup(t, s, batchDenom, ask.Denom, ask.Denom[1:], classId, start, end, creditType)
+
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(any interface{}, p *ecocredit.Params) {
+		p.CreditTypes = []*ecocredit.CreditType{&creditType}
+	}).Times(1)
+
+	future := time.Date(2077, 1, 1, 1, 1, 1, 1, time.Local)
+	middle := time.Date(2022, 1, 1, 1, 1, 1, 1, time.Local)
+	past := time.Date(1970, 1, 1, 1, 1, 1, 1, time.Local)
+
+	_, err := s.k.Sell(s.ctx, &marketplace.MsgSell{
+		Owner: s.addr.String(),
+		Orders: []*marketplace.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "5.22", AskPrice: &ask, DisableAutoRetire: false, Expiration: &past},
+		},
+	})
+	assert.NilError(t, err)
+
+	s.sdkCtx = s.sdkCtx.WithBlockTime(middle)
+	s.ctx = sdk.WrapSDKContext(s.sdkCtx)
+
+	// should work with future time
+	_, err = s.k.UpdateSellOrders(s.ctx, &marketplace.MsgUpdateSellOrders{
+		Owner: s.addr.String(),
+		Updates: []*marketplace.MsgUpdateSellOrders_Update{
+			{SellOrderId: 1, NewExpiration: &future},
+		},
+	})
+	assert.NilError(t, err)
+
```

Comment:
Maybe we should add a test case for `middle` as well.

Reply:
added. although now im confused, is it right that we allow expirations that are the same as the current block time? 🤔

Reply:
The sell order would be created and then it would never expire because we expire in `BeginBlocker`. It sounds like a bug to me and we should either update here or track with a new issue.

Reply:
wouldn't it just expire in the next block? iirc we prune from 0 to currentBlockTime, so it seems like it would just get immediately deleted upon the next block. ill update here to ensure the expiry time is _after_ the current block

Reply:
Ok, great. Seems odd that we allow for an expiration that is the same as the current time but also not an issue and not necessary to update.